proof/verifier.go

```diff
@@ -221,30 +238,203 @@ func (p *Proof) verifyExclusionProofs() (*commitment.TapCommitmentVersion,
 		return nil, ErrMissingExclusionProofs
 	}
 
-	// If there were no commitments in any exclusion proofs, then there is
-	// no version to return.
+	// If no commitments in proofs, no version to return.
 	if len(commitVersions) == 0 {
 		return nil, nil
 	}
 
-	// All ExclusionProofs must have similar versions.
-	firstCommitVersion := maps.Values(commitVersions)[0]
-	for outputIdx, commitVersion := range commitVersions {
-		outputCommitVersion := commitVersion
-		if !commitment.IsSimilarTapCommitmentVersion(
-			&firstCommitVersion, &outputCommitVersion,
-		) {
+	// All proofs must have similar versions.
+	firstVersion := maps.Values(commitVersions)[0][0]
+	return p.verifySTXOVersions(firstVersion, commitVersions)
+}
+
+// verifyV2ExclusionProofs verifies all version 2 exclusion proofs.
+func (p *Proof) verifyV2ExclusionProofs(
+	p2trOutputs map[uint32]fn.Set[asset.SerializedKey]) (
+	*commitment.TapCommitmentVersion, error) {
+
+	commitVersions := make(map[uint32][]commitment.TapCommitmentVersion)
+
+	// Collect the STXOs from the new asset.
+	stxoAssets, err := asset.CollectSTXO(&p.Asset)
+	if err != nil {
+		return nil, fmt.Errorf("error collecting STXO assets: %w", err)
+	}
+
+	// Map STXOs by serialized key.
+	assetMap := make(map[asset.SerializedKey]*asset.Asset)
+	for idx := range stxoAssets {
+		stxoAsset := stxoAssets[idx].(*asset.Asset)
+		key := asset.ToSerialized(stxoAsset.ScriptKey.PubKey)
+		assetMap[key] = stxoAsset
+		for outIdx := range p2trOutputs {
+			p2trOutputs[outIdx].Add(key)
+		}
+	}
+
+	for idx := range p.ExclusionProofs {
+		baseProof := p.ExclusionProofs[idx]
+		if baseProof.CommitmentProof == nil {
+			err := p.handleBasicExclusionProof(&baseProof,
```

Comment:
Same line wrapping comment here, won't add others as to not spam things.

Reply:
/me shakes fist at Aider